### PR TITLE
Add MDC2 support

### PR DESCRIPTION
**Pull request by damon-kwok**
Added `MDC2` hash function, digest and test cases.

### CODE REVIEW

File: crypto/_test.pony [hunk 1]
```diff
@@ -8,6 +8,7 @@ actor Main is TestList
     test(_TestConstantTimeCompare)
     test(_TestMD4)
     test(_TestMD5)
+    test(_TestMDC2)
```
Comment (SeanTAllen) [inline crypto/_test.pony:11]: can you move this before the _TestMD4 so it alphabetical?

File: crypto/_test.pony [hunk 2]
```diff
@@ -120,6 +129,13 @@ class iso _TestDigest is UnitTest
       "94af09c09bb9bb7b5c94fec6e6121482",
       ToHexString(md5.final()))
 
+    let mdc2 = Digest.mdc2()
```
Comment (SeanTAllen) [inline crypto/_test.pony:132]: Please move this before the mdc4 test so that it is alphabetical.

Comment (damon-kwok) [reply crypto/_test.pony:132]: NP. But I was thinking about whether I should give up this PR. OpenSSL sets a lot of compilation options, and each release handles it differently.

**Comment by damon-kwok** [comment]
All hashes support digests, which is no problem. But some hash algorithms are not implemented by openssl, such as shake, blake2b blake2s, they are implemented by other organizations, so the hash function does not appear in openssl, but they can be used in the abstract mode in the `EVP` framework.

File: crypto/_test.pony (hunk 2 above)
Comment (SeanTAllen) [reply crypto/_test.pony]: What compilation options? Are you saying that mdc2 isn't available sometimes based on the particular release and then is and then isn't again?

**Comment by SeanTAllen** [comment]
> All hashes support digests, which is no problem. But some hash algorithms are not implemented by openssl, such as shake, blake2b blake2s, they are implemented by other organizations, so the hash function does not appear in openssl, but they can be used in the abstract mode in the EVP framework.For those hash functions, we need to find the corresponding function name and call it directly.

Sorry, I don't understand. The idea is that you then need to link additional libraries to get some of the hash functions?

**Comment by damon-kwok** [comment]
The mdc2 is included in all openssl versions, but users can do this when building:
```
#define OPENSSL_NO_MDC2
```
or 
```cmake
-DOPENSSL_NO_MDC2=off
```
https://github.com/openssl/openssl/blob/dd76b90ef6cf9bd344c9a6cd0de536a734d1b6a3/include/openssl/evp.h#L800

**Comment by SeanTAllen** [comment]
They can do that with any of the definitions it appears. It seems to me if it is on by default, then it is reasonable for us to consider supporting it.

**Comment by SeanTAllen** [comment]
Is CI failing because the alpine distro version is off?

**Comment by damon-kwok** [comment]
The openssl in our test image does not contain mdc2, they are enabled by default in popular linux.
https://manpages.debian.org/stretch/openssl/mdc2.1ssl.en.html

**Comment by SeanTAllen** [comment]
What defines "popular Linux"? Alpine is quite popular for containers.

**Comment by damon-kwok** [comment]
Maybe we need to support hashing and encryption at runtime or at the standard library level. They are indispensable in a production environment.

**Comment by damon-kwok** [comment]
I think the container always chooses to minimize installation, and users need to enable it when needed.

**Comment by SeanTAllen** [comment]
How does moving to the standard library help? Seems like that is just moving the problem around.

I find "indispensable" to be a bit suspect. Supporting major algos seems important. However, alpine is doing fine with some algos turned off, so, I'm not sure how "indispensable" is being used here. We can't support every algo. Saying "popular" and "indispensable" doesn't really help come up with a rubric for what should be supported.

**Comment by SeanTAllen** [comment]
> I think the container always chooses to minimize installation, and users need to enable it when needed.

If this was true then every algo for openssl and libressl that can be turned on or off would be off. That is not the case as our CI shows.

**Comment by damon-kwok** [comment]
I understand that many production line environments use CentOS Image. Alpine Linux is an independent, non-commercial, general-purpose Linux distribution designed for advanced users who understand security, simplicity, and resource efficiency.

**Comment by damon-kwok** [comment]
> If this was true then every algo for openssl and libressl that can be turned on or off would be off. That is not the case as our CI shows.
For ponylang/crypto, we can choose to do so.

**Comment by SeanTAllen** [comment]
>> If this was true then every algo for openssl and libressl that can be turned on or off would be off. That is not the case as our CI shows.

> For ponylang/crypto, we can choose to do so.

Choice to do what?

**Comment by damon-kwok** [comment]
All algorithms of openssl and libressl that can be turned on or off will be turned off.

**Comment by SeanTAllen** [comment]
Sorry @damon-kwok I don't understand your point.

**Comment by damon-kwok** [comment]
https://github.com/BLAKE3-team/BLAKE3
Blake3 is released. But unless we can use openssl3, we cannot easily support it. Perhaps we should bypass the dependence on operating system libraries and integrate them directly.

**Comment by damon-kwok** [comment]
In general, what I want to express is that we can build popular algorithms into pony and use the latest stable version in time, because they usually have better security and high performance.

**Comment by SeanTAllen** [comment]
what does "integrate them directly" mean?

**Comment by SeanTAllen** [comment]
This can't be merged until:

- https://github.com/ponylang/shared-docker/issues/9
- https://github.com/ponylang/shared-docker/issues/10

are done.

**Comment by damon-kwok** [comment]
cool!

**Comment by damon-kwok** [comment]
> what does "integrate them directly" mean?

It probably means what @mfelsche  mentioned:
> What i meant was is providing e.g. openssl sources in the source tree of this library and compile them before actually compiling the pony sources, and automatically link them against this provided and compiled library.

**Comment by SeanTAllen** [comment]
https://github.com/ponylang/crypto/pull/41#issuecomment-653901887 still need to be completed before this can be rebased, tested, and merged.

**Comment by SeanTAllen** [comment]
libressl has no support for MDC2 that I can find.

**Comment by damon-kwok** [comment]
This means that MDC2 support needs to be abandoned.

**Pull request closed without merging**